plan/physical_plan_builder.go

```diff
@@ -175,27 +189,42 @@ func (ds *DataSource) findBestTask(prop *requiredProp) (task, error) {
 		return nil, nil
 	}
 
-	t := ds.getTask(prop)
+	t = ds.getTask(prop)
 	if t != nil {
 		return t, nil
 	}
-	t, err := ds.tryToGetDualTask()
+
+	// If prop.enforced is true, the prop.cols need to be set nil for ds.findBestTask.
+	// Before function return, reset it for enforcing task prop and storing map<prop,task>.
+	oldPropCols := prop.cols
+	if prop.enforced {
+		prop.cols = []*expression.Column{}
+	}
+	defer func() {
+		if err != nil {
+			return
+		}
+		if prop.enforced {
+			prop.cols = oldPropCols
+			t = prop.enforceProperty(t, ds.basePlan.ctx)
+		}
+		ds.storeTask(prop, t)
+	}()
+
+	t, err = ds.tryToGetDualTask()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	if t != nil {
```

Comment:
how about merge this `if` with the `if` statement located in line 215 together to make code simpler:
```go
if err != nil || t != nil {
    return t, errors.Trace(err)
}
```

Reply:
I couldn't understand why the **two** `if` can merge.